In [1]:
# This Python 3 environment comes with many helpful analytics libraries installed
# It is defined by the kaggle/python Docker image: https://github.com/kaggle/docker-python
# For example, here's several helpful packages to load

import numpy as np # linear algebra
import pandas as pd # data processing, CSV file I/O (e.g. pd.read_csv)

# Input data files are available in the read-only "../input/" directory
# For example, running this (by clicking run or pressing Shift+Enter) will list all files under the input directory

import os
for dirname, _, filenames in os.walk('/kaggle/input'):
    for filename in filenames:
        print(os.path.join(dirname, filename))

# You can write up to 20GB to the current directory (/kaggle/working/) that gets preserved as output when you create a version using "Save & Run All" 
# You can also write temporary files to /kaggle/temp/, but they won't be saved outside of the current session

/kaggle/input/2022-ml-midterm-p2/submit.csv
/kaggle/input/2022-ml-midterm-p2/train.csv
/kaggle/input/2022-ml-midterm-p2/test.csv


In [2]:
import random
import os
seed = 42
random.seed(seed)
np.random.seed(seed)
os.environ["PYTHONHASHSEED"] = str(seed)

In [3]:
sample = pd.read_csv('/kaggle/input/2022-ml-midterm-p2/submit.csv')
train = pd.read_csv('/kaggle/input/2022-ml-midterm-p2/train.csv')
test = pd.read_csv('/kaggle/input/2022-ml-midterm-p2/test.csv')

데이터를 불러오기 위해서 pd.read_csv() 함수를 사용하였다.

In [4]:
train

index  0  1   2   3   4   5  6  7  8  ...  55  56  57  58  59  60  61  \
0         0  0  0   5  13   9   1  0  0  0  ...   0   0   0   6  13  10   0   
1         1  0  0   0  12  13   5  0  0  0  ...   0   0   0   0  11  16  10   
2         2  0  0   0   4  15  12  0  0  0  ...   0   0   0   0   3  11  16   
3         3  0  0   7  15  13   1  0  0  0  ...   0   0   0   7  13  13   9   
4         4  0  0   0   1  11   0  0  0  0  ...   0   0   0   0   2  16   4   
...     ... .. ..  ..  ..  ..  .. .. .. ..  ...  ..  ..  ..  ..  ..  ..  ..   
1432   1792  0  0   4  10  13   6  0  0  0  ...   0   0   0   2  14  15   9   
1433   1793  0  0   6  16  13  11  1  0  0  ...   0   0   0   6  16  14   6   
1434   1794  0  0   1  11  15   1  0  0  0  ...   0   0   0   2   9  13   6   
1435   1795  0  0   2  10   7   0  0  0  0  ...   0   0   0   5  12  16  12   
1436   1796  0  0  10  14   8   1  0  0  0  ...   0   0   1   8  12  14  12   

      62  63  label  
0      0   0      0  
1      0   0      1  
2      9   0      2  
3      0   0      3  
4      0   0      4  
...   ..  ..    ...  
1432   0   0      9  
1433   0   0      0  
1434   0   0      8  
1435   0   0      9  
1436   1   0      8  

[1437 rows x 66 columns]

In [5]:
X_train = train.drop(['index','label'],axis=1)
y_train = train['label']
X_test = test.drop(['index'],axis=1)

데이터를 살펴보니 'index'가 공통적으로 훈련데이터와 학습데이터에 들어있어서 제거해주었고, 훈련데이터에서는 학습을 하기위해서 'label'라는 라벨을 따로 떼서 다른 변수에 담아주었다.

In [6]:
X_train

0  1   2   3   4   5  6  7  8  9  ...  54  55  56  57  58  59  60  61  \
0     0  0   5  13   9   1  0  0  0  0  ...   0   0   0   0   6  13  10   0   
1     0  0   0  12  13   5  0  0  0  0  ...   0   0   0   0   0  11  16  10   
2     0  0   0   4  15  12  0  0  0  0  ...   5   0   0   0   0   3  11  16   
3     0  0   7  15  13   1  0  0  0  8  ...   9   0   0   0   7  13  13   9   
4     0  0   0   1  11   0  0  0  0  0  ...   0   0   0   0   0   2  16   4   
...  .. ..  ..  ..  ..  .. .. .. .. ..  ...  ..  ..  ..  ..  ..  ..  ..  ..   
1432  0  0   4  10  13   6  0  0  0  1  ...   4   0   0   0   2  14  15   9   
1433  0  0   6  16  13  11  1  0  0  0  ...   1   0   0   0   6  16  14   6   
1434  0  0   1  11  15   1  0  0  0  0  ...   0   0   0   0   2   9  13   6   
1435  0  0   2  10   7   0  0  0  0  0  ...   2   0   0   0   5  12  16  12   
1436  0  0  10  14   8   1  0  0  0  2  ...   8   0   0   1   8  12  14  12   

      62  63  
0      0   0  
1      0   0  
2      9   0  
3      0   0  
4      0   0  
...   ..  ..  
1432   0   0  
1433   0   0  
1434   0   0  
1435   0   0  
1436   1   0  

[1437 rows x 64 columns]

In [7]:
from sklearn.neighbors import KNeighborsClassifier
model2 = KNeighborsClassifier().fit(X_train,y_train)
y_test = model2.predict(X_test)

해당 문제는 손글씨 숫자 인식을 하는 문제로 분류 문제에 해당하기에 먼저 분류모델을 만들어야한다고 생각했고 학습데이터의 수가 1436개 정도로 많으므로 KNeighborsClassifier을 선언하고 학습을 시켰다.

In [8]:
sample['label'] = y_test
sample.to_csv('submit.csv',index=False)

모델을 통해서 에측한 값을 csv에 저장하기 위해서 .to_csv라는 함수를 사용했다.

In [9]:
from sklearn.ensemble import RandomForestClassifier
model1 = RandomForestClassifier().fit(X_train,y_train)
y_test1 = model1.predict(X_test)

해당 문제는 분류 문제로 분류모델을 만들어야겠다고 생각했고, 학습의 데이터의 수가 많고 설명보다는 예측을 필요로 하는 문제이기에 DecisionTreeClassifier보다는 RandomForestClassifier가 더 적합하다고 생각하여 RandomForestClassifier을 선언하고 학습을 시켰다

In [10]:
#sample['label'] = y_test1
#sample.to_csv('submit1.csv',index=False)

모델을 통해서 에측한 값을 csv에 저장하기 위해서 .to_csv라는 함수를 사용했다.

KNeighborsClassifier으로 모델을 만들고 파라미터를 default 값으로 설정후 캐글에 제출해보니 0.98611 가 나왔고 RandomForestClassifier로 모델을 만들고 파라미터를 default 값으로 설정후 캐글에 제출해보니 0.97500 가 나왔다. 이를 통해서 분석을 해보자면 해당 데이터에서는 우선 학습데이터가 많고 데이터의 분산이 그리 크지 않아서 KNeighborsClassifier의 성능이 더 높을 것으로 예상을 하였고 이에 캐글의 점수도 KNeighborsClassifier의 성능이 더 좋다고 나왔다. KNeighborsClassifier와 RandomForestClassifier의 성능 차이는 0.01점 정도로 크게 차이가 나지 않지만 해당 데이터에서는 우선 데이터가 많고 데이터의 분산이 크지 않아서 KNeighborsClassifier가 조금 더 성능이 좋게 나온 것으로 예상합니다.